python/packages/ai/teams/ai/ai.py

```diff
@@ -76,3 +100,221 @@ def __call__(func: Callable[[TurnContext, StateT, Any, str], Awaitable[bool]]):
             return func
 
         return __call__
+
+    def review_prompt(
+        self, func: Callable[[TurnContext, StateT, PromptTemplate], Awaitable[Optional[Plan]]]
+    ):
+        """
+        Reviews user input before it's sent to the planner.
+
+        ```python
+        # Use this method as a decorator
+        @app.ai.review_prompt
+        async def on_review_prompt(context: TurnContext, state: TurnState, prompt: PromptTemplate):
+            return None
+
+        # Pass a function to this method
+        app.ai.review_prompt(on_review_prompt)
+        ```
+
+        Args:
+            context (TurnContext): The turn context.
+            state (StateT): The turn state.
+            prompt (PromptTemplate): The prompt.
+
+        Returns:
+            Optional[Plan]: Returns None if the moderator approves the input,
+            otherwise returns a new plan to bypass the planner and redirects to
+            a new set of actions. Typically the moderator will return a new plan
+            with a single DO command that calls `AI.FlaggedInputActionName`
+            to flag the input for review.
+        """
+
+        self._review_prompt = func
+        return func
+
+    def review_plan(self, func: Callable[[TurnContext, StateT, Plan], Awaitable[Plan]]):
+        """
+        Reviews a plan generated by the planner before its executed.
+
+        ```python
+        # Use this method as a decorator
+        @app.ai.review_plan
+        async def on_review_plan(context: TurnContext, state: TurnState, plan: Plan):
+            return plan
+
+        # Pass a function to this method
+        app.ai.review_plan(on_review_plan)
+        ```
+
+        Args:
+            context (TurnContext): The turn context
+            state (StateT): The turn state
+            plan (Plan): The plan to review.
+
+        Returns:
+            Plan: Returns the same plan if the moderator approves the plan,
+            otherwise returns a new plan that redirects to a new set of actions.
+            Typically the moderator will return a new plan with a single DO command
+            that calls `AI.FlaggedOutputActionName` to flag the output for review.
+        """
+
+        self._review_plan = func
+        return func
+
+    def function(self, name: Optional[str] = None, allow_overrides=False):
```

Comment:
Same with another comment, since prompt manager is part of planner, maybe we shouldn't add the entrypoint to prompt manager here? Just `planner.add_function()` is fine.

Reply:
I think since we have chosen to use decorators anything the user overrides should be accessible via a decorator. This provides a simple decorator for users to register functions the same way they would register an action, so the syntax is consistent.

Reply:
I‘m neutral to this. Can you create tasks for JS/C# to add this? It should be doable in all SDKs and we can drive this if thinking it's valuable to users.